### PR TITLE
Distributed Deep Learning image end to end tutorial

### PR DESCRIPTION
**Pull request by jfomhover**
# PR into Azure/azureml-examples

This PR adds an end to end jupyter notebook tutorial for distributed pytorch.

Here are the learning objectives of this tutorial:
- quickly implement basic commands for data preparation
- assemble a pipeline with custom data preparation (python) scripts
- test and run a multi-node multi-gpu pytorch job
~~- register and deploy a pytorch model~~
- use mlflow to analyze your metrics

The tutorial relies on a distributed pytorch training script for image classification using coco dataset. This code implements the following:
- multi gpu, multi node pytorch distributed training
- reporting metrics using mlflow
- supports pytorch profiler reporting as markdown/tensorboard in mlflow

Future changes (known limitations):
- implement deployement
- implement a short "distributed benchmark" (waiting for parameters->runsettings implementation)
- implement checkpointing

### CODE REVIEW

File: sdk/jobs/e2e-dl-cnn-distributed/src/pytorch_dl_train/model.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    )
+    if model_arch in MODEL_ARCH_LIST:
+        model = getattr(models, model_arch)(pretrained=pretrained)
+    else:
+        raise NotImplementedError(
+            f"model_arch={model_arch} is not implemented in torchvision model zoo."
+        )
+
+    if model_arch == "resnet18":
+        model.fc = nn.Linear(model.fc.in_features, output_dimension)
```
Comment (jfomhover) [inline sdk/jobs/e2e-dl-cnn-distributed/src/pytorch_dl_train/model.py:71]: this is wrong and I need to fix that

File: sdk/jobs/e2e-dl-cnn-distributed/e2e-object-classification-distributed-pytorch.ipynb [hunk 2]
```diff
@@ -0,0 +1,736 @@
...
+    "    # we set the name of the compute target for this training job\n",
+    "    training_step.compute = \"gpu-cluster\"\n",
+    "\n",
+    "    # use process_count_per_instance to parallelize on multiple gpus\n",
+    "    training_step.distribution.process_count_per_instance = (\n",
+    "        1  # set to number of gpus on instance\n",
+    "    )\n",
+    "\n",
+    "    # use instance_count to increase the number of nodes (machines)\n",
+    "    training_step.resources.instance_count = 1\n",
```
[the remark this answers is not in the document]
Comment (jfomhover) [reply sdk/jobs/e2e-dl-cnn-distributed/e2e-object-classification-distributed-pytorch.ipynb:474]: I'll have to add it to the backlog, we're waiting for a couple sdk features to be able to implement this in a good fashion (in particular, I'd need to be able to set number of nodes from a pipeline parameter.

File: sdk/jobs/e2e-dl-cnn-distributed/src/pytorch_dl_train/model.py (hunk 1 above)
Comment (jfomhover) [reply sdk/jobs/e2e-dl-cnn-distributed/src/pytorch_dl_train/model.py]: ok I'd rather implement multiclass later, let's merge this one as is and build upon it iteratively

File: sdk/jobs/e2e-dl-cnn-distributed/e2e-object-classification-distributed-pytorch.ipynb [hunk 3]
```diff
@@ -0,0 +1,736 @@
...
+   "execution_count": null,
+   "id": "1c79abd9",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "from azure.ml import dsl\n",
+    "\n",
+    "# the dsl decorator tells the sdk that we are defining an AML pipeline\n",
+    "@dsl.pipeline(\n",
+    "    compute=\"cpu-cluster\",\n",
```
[the remark this answers is not in the document]
Comment (jfomhover) [reply sdk/jobs/e2e-dl-cnn-distributed/e2e-object-classification-distributed-pytorch.ipynb]: I added a section to "Provision the requires resources" which should bring clarity here

File: sdk/jobs/e2e-dl-cnn-distributed/e2e-object-classification-distributed-pytorch.ipynb [hunk 4]
```diff
@@ -0,0 +1,736 @@
...
+    "    environment=\"AzureML-sklearn-0.24-ubuntu18.04-py37-cpu:9\",\n",
+    ")"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "00903f3a",
+   "metadata": {},
+   "source": [
+    "The component we just created can now be loaded as a component: a reusable step in a pipeline. "
```
[the remark this answers is not in the document]
Comment (jfomhover) [reply sdk/jobs/e2e-dl-cnn-distributed/e2e-object-classification-distributed-pytorch.ipynb]: Good point, I applied the rewording.

**Review by lostmygithubaccount — APPROVED** [review]

**Review by AliMahmoudzadeh — APPROVED** [review]

**Pull request merged**